{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13f85682",
   "metadata": {},
   "outputs": [],
   "source": [
    "####import neccessary modules######\n",
    "import csv #import csv so i could csvreader but i didnt actually use this\n",
    "import pandas as pd # pandas to convert my csv to a dataframe\n",
    "from sklearn.preprocessing import StandardScaler #for normalisation\n",
    "import numpy as np \n",
    "from sklearn.decomposition import PCA # for PCA\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import manifold\n",
    "from sklearn.manifold import TSNE\n",
    "import mpl_toolkits.mplot3d\n",
    "import seaborn as sns"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4be17e87",
   "metadata": {},
   "source": [
    "Great! trying this out!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a05a8d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####read in gene expression csv as a dataframe, index so it reads in row names\n",
    "expression = pd.read_csv('/home/alanah/Downloads/Expression_Public_22Q4.csv', index_col=0) \n",
    "\n",
    "#######extracting the column names and extracting rwo names\n",
    "column_names = list(expression.columns.values) \n",
    "row_names = list(expression.index.values) \n",
    "\n",
    "####index all the values from the expression datafram, not sure if this is useful\n",
    "x = expression.iloc[:, ].values \n",
    "\n",
    "##### normalizing the features\n",
    "#x = StandardScaler().fit_transform(x) \n",
    "\n",
    "#####whats the dimensions of the data, check mean and stadard deviation of normalised data\n",
    "print(x.shape) \n",
    "print(np.mean(x),np.std(x)) \n",
    "\n",
    "######make a PCA q with 2 principal components \n",
    "pca_expression = PCA(n_components=2) \n",
    "pca_expression.fit_transform(np.transpose(x))\n",
    "\n",
    "#### look at how much variance is explained by the PCs\n",
    "print(pca_expression.explained_variance_ratio_)\n",
    "\n",
    "######plot the PCA as a scatter plot with a component on eah axis , [0, : ] -> the fir\n",
    "plt.scatter(pca_expression.components_[0,:], pca_expression.components_[1,:])  \n",
    "plt.title('DepMap Gene Expression Principal Component Analysis')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.show() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5787290d",
   "metadata": {},
   "outputs": [],
   "source": [
    "######read in gene expression csv as a dataframe, index so it reads in row names\n",
    "mutation = pd.read_csv('/home/alanah/Downloads/Damaging_Mutations.csv', index_col=0) \n",
    "\n",
    "\n",
    "## comment out this line to only include heterozygous damaging mutations, comment in to include both heterozygous and homozygous mutations\n",
    "#mutation = mutation.replace(2,1) \n",
    "\n",
    "####extracting the column names and extracting row names\n",
    "column_names = list(mutation.columns.values) \n",
    "row_names_mut = list(mutation.index.values) \n",
    "\n",
    "##### index all the values from the expression datafram, not sure if this is useful\n",
    "x_mut = mutation.iloc[:, ].values \n",
    "print(x_mut)\n",
    "#x_mut = StandardScaler().fit_transform(x_mut) \n",
    "\n",
    "###### normalizing the features\n",
    "#x_mut = StandardScaler().fit_transform(x_mut) \n",
    "\n",
    "####whats the dimensions of the data, check mean and stadard deviation of normalised data\n",
    "#print(x.shape) \n",
    "#print(np.mean(x),np.std(x)) \n",
    "#print(type(x))\n",
    "\n",
    "#### make a PCA q with 2 principal components \n",
    "pca_mutation = PCA(n_components=2) \n",
    "pca_mutation.fit_transform(np.transpose(x_mut))\n",
    "print(pca_mutation.explained_variance_ratio_)\n",
    "\n",
    "###### plot the PCA as a scatter plot with a component on eah axis\n",
    "plt.scatter(pca_mutation.components_[0,:] ,pca_mutation.components_[1,:]) \n",
    "plt.title('DepMap Mutation Principal Component Analysis')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.show() \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a201eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### sum up the amount of mutations per gene\n",
    "sum_genes = np.sum(x_mut, axis = 0)\n",
    "\n",
    "#####sort the amount of mutations per gene in descending order\n",
    "top_genes = np.argsort(sum_genes)\n",
    "\n",
    "#####get the indexes of the top five most mutated genes in one list together and individually\n",
    "top_five = top_genes[-5:, ]\n",
    "gene_one = (top_genes[-1,])\n",
    "gene_two = (top_genes[-2,])\n",
    "gene_three = (top_genes[-3,])\n",
    "gene_four = (top_genes[-4,])\n",
    "gene_five = (top_genes[-1,])\n",
    "\n",
    "####### find the names of the most mutated genes\n",
    "top_five_names = list()\n",
    "for i in top_five:\n",
    "    top_five_names.append(mutation.columns[i])\n",
    "\n",
    "print(top_five_names)\n",
    "#print(mutation.columns[gene_five])\n",
    "print((np.where(x_mut[:,gene_one]==1)))\n",
    "\n",
    "#for i in top_five:\n",
    "  # plt.scatter(pca_mutation.components_[0,np.where(x_mut[:,[i]]==0)] ,pca_mutation.components_#[1,np.where(x_mut[:,[i]]==0)]) \n",
    "  # plt.scatter(pca_mutation.components_[0,np.where(x_mut[:,[i]]==1)],pca_mutation.components_[1,np.where(x_mut[:,[i]]==1)])\n",
    "  # i = i + 1\n",
    " \n",
    "\n",
    "# sum the number of times adamaging mutation is found in the top gene for all cell lines \n",
    "\n",
    "\n",
    "######do the PCA\n",
    "# sum up the number of times no mutation is found for the top mutated gene\n",
    "color_1= ('tab:blue')\n",
    "color_2= ('tab:orange')\n",
    "plt.scatter(pca_mutation.components_[0,np.where(x_mut[:,gene_five]==0)] ,pca_mutation.components_[1,np.where(x_mut[:,gene_five]==0)], c = color_1 ,label='None') \n",
    "\n",
    "# sum the number of times adamaging mutation is found in the top gene for all cell lines \n",
    "plt.scatter(pca_mutation.components_[0,np.where(x_mut[:,gene_five]==1)],pca_mutation.components_[1,np.where(x_mut[:,gene_five]==1)], c = color_2, label = 'Damaging' )\n",
    "\n",
    "plt.title('Principal Component Analysis of no vs damaging mutation in CYP2D6')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.legend()\n",
    "\n",
    "\n",
    "\n",
    "plt.show() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ce2ea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### find the lengths of the rows of the expression and mutation dataset\n",
    "print(len(row_names))\n",
    "print(len(row_names_mut))\n",
    "\n",
    "\n",
    "#### create a for loop to get the overlap between the two lists\n",
    "overlap_exp  = list()\n",
    "overlap_mut = list()\n",
    "for i in range(len(row_names)):     # for each variable in the range of the full length of the row_names list\n",
    "    if row_names_mut.count(row_names[i])>0: # if the row name in row_names is counted at least once\n",
    "        overlap_exp.append(i) # append that row_name index to the overlap_exp list\n",
    "        overlap_mut.append(row_names_mut.index(row_names[i])) # append the index where the row name from row_names appears in row_names_mut\n",
    "\n",
    "\n",
    "####### find the length of overlap_exp and overlap_mut, chek they are the same\n",
    "print(len(overlap_exp))  \n",
    "print(len(overlap_mut))\n",
    "len(overlap_exp) == len(overlap_mut)\n",
    "\n",
    "######### create new dataframesw containing from the main ones, containing only the rows that match those from the overlap lists. .iloc: matching indices to indices as the overlap lists contain the row numbers (the indices) ########\n",
    "\n",
    "joinable_exp = expression.iloc[overlap_exp]  \n",
    "print(joinable_exp)\n",
    "\n",
    "joinable_mut = mutation.iloc[overlap_mut]\n",
    "print(len(joinable_mut))\n",
    "\n",
    "\n",
    "#### sort the joinable datasets to be inascending order so now they not only contain the same row names but in the same order.\n",
    "joinable_exp = joinable_exp.sort_index(ascending=True)\n",
    "joinable_mut = joinable_mut.sort_index(ascending=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f5ad563",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### index all the values from the datafram, make PCA with 2 components, transform and transpose the data for bothe the datasets\n",
    "join_exp = joinable_exp.iloc[:, ].values\n",
    "#join_exp = StandardScaler().fit_transform(join_exp) \n",
    "pca_joinexp =PCA(n_components=2)\n",
    "pca_joinexp.fit_transform(np.transpose(join_exp))\n",
    "\n",
    "join_mut = joinable_mut.iloc[:, ].values\n",
    "#join_mut = StandardScaler().fit_transform(join_mut) \n",
    "pca_joinmut =PCA(n_components=2)\n",
    "pca_joinmut.fit_transform(np.transpose(join_mut))\n",
    "\n",
    "\n",
    "##### find the variance explained by them\n",
    "print(pca_joinmut.explained_variance_ratio_)\n",
    "print(pca_joinexp.explained_variance_ratio_)\n",
    "\n",
    "##### find the index of a specfic gene\n",
    "print(joinable_exp.columns.get_loc('TP53'))\n",
    "\n",
    "##### plot the PCA as a scatter plot with a the PC! from the mutation dataset and PC1 from the expression datasetc\n",
    "plt.scatter(pca_joinmut.components_[0,:], pca_joinexp.components_[0,:]) \n",
    "plt.title('Principal Component Analysis of PC1 of the expression data and PC1 of the mutation data')\n",
    "plt.xlabel('PC1 from mutation ')\n",
    "plt.ylabel('PC1 from expression')\n",
    "\n",
    "\n",
    "plt.show() # show PCA plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46af56ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### indexing all the columns where gene one has a damaging mutation (row indexes)\n",
    "mut_gene_one =np.where(x_mut[:,gene_one]==1)\n",
    "\n",
    "#####  getting all the cell lines names ie the row names for the cell lines with a damaging mutation in gene one\n",
    "cell_lines = list(mutation.index[mut_gene_one])\n",
    "\n",
    "\n",
    "###### for loop to get the row indexes where \n",
    "mut_celllines = list() #list of cellines (rows) where the damaging mutation appear\n",
    "no_mut_celllines = list() #list of cell lines (rows) where no mutation occurs\n",
    "for i in range(len(row_names)):     # for each variable in the range of the full length of the row_names list\n",
    "    if cell_lines.count(row_names[i])>0: # if the row name in row_names is counted at least once\n",
    "        mut_celllines.append(i)\n",
    "    else: # append that row_name index to the overlap_exp list\n",
    "        no_mut_celllines.append(i) # append the index where the row name from row_names appears in row_names_mut\n",
    "\n",
    "\n",
    "print(len(mut_celllines))\n",
    "print(len(no_mut_celllines))\n",
    "\n",
    "###### new variabe with just the rows from the expression data for the cell lines with a damaging mutation and the inverse. Also doing the PCA bits\n",
    "geneone_mutexp = expression.iloc[mut_celllines] \n",
    "geneone_nomutexp = expression.iloc[no_mut_celllines]  \n",
    "\n",
    "pca_e =PCA(n_components=2)\n",
    "pca_e.fit_transform(np.transpose(expression.iloc[:,].values))\n",
    "\n",
    "plt.scatter(pca_e.components_[0,no_mut_celllines], pca_e.components_[1,no_mut_celllines], c=color_1, label='None')\n",
    "plt.scatter(pca_e.components_[0,mut_celllines], pca_e.components_[1,mut_celllines], c=color_2, label = 'Damaging')\n",
    "plt.title('Principal Component Analysis of expression for CYP2D6 based on the presence of a damaging mutation')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.legend()\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#print(np.where(x[a,:]))\n",
    "#joinable_mut = mutation.iloc[overlap_mut]\n",
    "#print(len(joinable_mut))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46af56ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### indexing all the columns where gene one has a damaging mutation (row indexes)\n",
    "mut_gene_one =np.where(x_mut[:,gene_one]==1)\n",
    "\n",
    "#####  getting all the cell lines names ie the row names for the cell lines with a damaging mutation in gene one\n",
    "cell_lines = list(mutation.index[mut_gene_one])\n",
    "\n",
    "\n",
    "###### for loop to get the row indexes where \n",
    "mut_celllines = list() #list of cellines (rows) where the damaging mutation appear\n",
    "no_mut_celllines = list() #list of cell lines (rows) where no mutation occurs\n",
    "for i in range(len(row_names)):     # for each variable in the range of the full length of the row_names list\n",
    "    if cell_lines.count(row_names[i])>0: # if the row name in row_names is counted at least once in the list of cell lines \n",
    "        mut_celllines.append(i)  # append that row_name index to the mut_celllines list\n",
    "    else:\n",
    "        no_mut_celllines.append(i) # append that row_name index to the no_mut_celllines list\n",
    "\n",
    "\n",
    "print(len(mut_celllines))\n",
    "print(len(no_mut_celllines))\n",
    "\n",
    "###### new variabe with just the rows from the expression data for the cell lines with a damaging mutation and the inverse. Also doing the PCA bits\n",
    "geneone_mutexp = expression.iloc[mut_celllines] \n",
    "geneone_nomutexp = expression.iloc[no_mut_celllines]  \n",
    "\n",
    "pca_mutexp =PCA(n_components=2)\n",
    "pca_mutexp.fit_transform(np.transpose(expression.iloc[:,].values))\n",
    "\n",
    "plt.scatter(pca_mutexp.components_[0,no_mut_celllines], pca_mutexp.components_[1,no_mut_celllines])\n",
    "plt.scatter(pca_mutexp.components_[0,mut_celllines], pca_mutexp.components_[1,mut_celllines])\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#print(np.where(x[a,:]))\n",
    "#joinable_mut = mutation.iloc[overlap_mut]\n",
    "#print(len(joinable_mut))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "8f664bc7b08e2611024fa64ff643512fe9a9d18971c14cf49d5a15ec0a97a790"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
